modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,651 @@
+                DummyType->TypeInfo.ParseProcedure = ParseProc;
+                DummyType->TypeInfo.OkayToCloseProcedure = OkayToCloseProc;
+                DummyType->TypeInfo.QueryNameProcedure = QueryNameProc;
+            }
+
+            /* Close the object type handle */
+            if (ObjectTypeHandle != NULL)
+            {
+                Status = ZwClose(ObjectTypeHandle);
+            }
+        }
+    }
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+}
+
+/*!
+ * @brief This function cleans up resources that were not cleaned up during the tests.
+ */
+static
+VOID
+ObInsert_Cleanup()
+{
+    NTSTATUS Status;
+
+    /* Make the directory object temporary (which is permanent) if the handle is valid */
+    if (!skip(DirectoryHandle != NULL, "No directory handle\n"))
+    {
+        Status = ZwMakeTemporaryObject(DirectoryHandle);
+        ok_eq_hex(Status, STATUS_SUCCESS);
+
+        /* Close the directory handle */
+        Status = ZwClose(DirectoryHandle);
+        ok_eq_hex(Status, STATUS_SUCCESS);
+    }
+}
+
+/*!
+ * @brief This function tests for successful object insertion.
+ */
+static
+VOID
+ObInsert_Success()
+{
+    NTSTATUS Status;
+    UNICODE_STRING ObjectName;
+    PDUMMY_TYPE Object;
+    HANDLE Handle;
+    POBJECT_HEADER ObjectHeader;
+
+    /* Initialize the object name */
+    RtlInitUnicodeString(&ObjectName,
+                         L"\\ObInsertTest\\DummyObject_Success");
+
+    /* Initialize object attributes with the name */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &ObjectName,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the object */
+    Status = ObCreateObject(KernelMode,
+                            DummyType,
+                            &DummyObjectAttributes,
+                            KernelMode,
+                            NULL,
+                            sizeof(DUMMY_TYPE),
+                            0,
+                            0,
+                            (PVOID *)&Object);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    /* Initialize the object data */
+    RtlZeroMemory(Object, sizeof(DUMMY_TYPE));
+    Object->Data = 123;
+
+    ObjectHeader = OBJECT_TO_OBJECT_HEADER(Object);
+
+    ok_eq_ulong(ObjectHeader->PointerCount, 1LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 0LU);
+
+    /* Insert the object */
+    Status = ObInsertObject(Object,
+                            NULL,
+                            STANDARD_RIGHTS_ALL,
+                            0,
+                            NULL,
+                            &Handle);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+    ok(Object != NULL, "Object is NULL\n");
+    ok(Handle != NULL, "Handle is NULL\n");
+
+    ok_eq_ulong(ObjectHeader->PointerCount, 2LU);
+    ok_eq_ulong(ObjectHeader->HandleCount, 1LU);
+
+    ObDereferenceObject(Object);
```

Comment:
Check this, in case this applies to your case:
https://git.reactos.org/?p=reactos.git;a=commit;h=fc6bf61e4c90caa3db244d137efdb873d4876e48